File: packages/core-php/src/CborDecoder.php
<?php

declare(strict_types=1);

namespace OpenPasskey;

class CborDecoder
{
    private const CBOR_MAJOR_UNSIGNED_INT = 0;
    private const CBOR_MAJOR_NEGATIVE_INT = 1;
    private const CBOR_MAJOR_BYTE_STRING = 2;
    private const CBOR_MAJOR_TEXT_STRING = 3;
    private const CBOR_MAJOR_ARRAY = 4;
    private const CBOR_MAJOR_MAP = 5;
    private const CBOR_MAJOR_TAG = 6;
    private const CBOR_MAJOR_FLOAT_SIMPLE = 7;

    public static function decode(string $data): mixed
    {
        $buf = new ByteBuffer($data);
        $offset = 0;
        $result = self::parseItem($buf, $offset);
        if ($offset !== $buf->getLength()) {
            throw new WebAuthnException('cbor_error', 'Unused bytes after data item');
        }
        return $result;
    }

    public static function decodeInPlace(string $data, int $startOffset, ?int &$endOffset = null): mixed
    {
        $buf = new ByteBuffer($data);
        $offset = $startOffset;
        $result = self::parseItem($buf, $offset);
        $endOffset = $offset;
        return $result;
    }

    private static function parseItem(ByteBuffer $buf, int &$offset): mixed
    {
        $first = $buf->getByteVal($offset++);
        $type = $first >> 5;
        $val = $first & 0b11111;

        if ($type === self::CBOR_MAJOR_FLOAT_SIMPLE) {
            return self::parseFloatSimple($val, $buf, $offset);
        }

        $val = self::parseExtraLength($val, $buf, $offset);

        return self::parseItemData($type, $val, $buf, $offset);
    }

    private static function parseFloatSimple(int $val, ByteBuffer $buf, int &$offset): mixed
    {
        switch ($val) {
            case 24:
                $val = $buf->getByteVal($offset++);
                return self::parseSimple($val);
            case 25:
                $f = $buf->getHalfFloatVal($offset);
                $offset += 2;
                return $f;
            case 26:
                $f = $buf->getFloatVal($offset);
                $offset += 4;
                return $f;
            case 27:
                $f = $buf->getDoubleVal($offset);
                $offset += 8;
                return $f;
            case 28:
            case 29:
            case 30:
                throw new WebAuthnException('cbor_error', 'Reserved value used');
            case 31:
                throw new WebAuthnException('cbor_error', 'Indefinite length is not supported');
        }

        return self::parseSimple($val);
    }

    private static function parseSimple(int $val): mixed
    {
        if ($val === 20) return false;
        if ($val === 21) return true;
        if ($val === 22) return null;
        throw new WebAuthnException('cbor_error', "Unsupported simple value {$val}");
    }

    private static function parseExtraLength(int $val, ByteBuffer $buf, int &$offset): int
    {
        switch ($val) {
            case 24:
                $val = $buf->getByteVal($offset++);
                break;
            case 25:
                $val = $buf->getUint16Val($offset);
                $offset += 2;
                break;
            case 26:
                $val = $buf->getUint32Val($offset);
                $offset += 4;
                break;
            case 27:
                $val = $buf->getUint64Val($offset);
                $offset += 8;
                break;
            case 28:
            case 29:
            case 30:
                throw new WebAuthnException('cbor_error', 'Reserved value used');
            case 31:
                throw new WebAuthnException('cbor_error', 'Indefinite length is not supported');
        }
        return $val;
    }

    private static function parseItemData(int $type, int $val, ByteBuffer $buf, int &$offset): mixed
    {
        switch ($type) {
            case self::CBOR_MAJOR_UNSIGNED_INT:
                return $val;
            case self::CBOR_MAJOR_NEGATIVE_INT:
                return -1 - $val;
            case self::CBOR_MAJOR_BYTE_STRING:
                $data = $buf->getBytes($offset, $val);
                $offset += $val;
                return $data;
            case self::CBOR_MAJOR_TEXT_STRING:
                $data = $buf->getBytes($offset, $val);
                $offset += $val;
                return $data;
            case self::CBOR_MAJOR_ARRAY:
                return self::parseArray($buf, $offset, $val);
            case self::CBOR_MAJOR_MAP:
                return self::parseMap($buf, $offset, $val);
            case self::CBOR_MAJOR_TAG:
                return self::parseItem($buf, $offset);
        }

        throw new WebAuthnException('cbor_error', "Unknown major type {$type}");
    }

    private static function parseMap(ByteBuffer $buf, int &$offset, int $count): array
    {
        $map = [];
        for ($i = 0; $i < $count; $i++) {
            $key = self::parseItem($buf, $offset);
            $value = self::parseItem($buf, $offset);
            if (!\is_int($key) && !\is_string($key)) {
                throw new WebAuthnException('cbor_error', 'Map keys must be strings or integers');
            }
            $map[$key] = $value;
        }
        return $map;
    }

    private static function parseArray(ByteBuffer $buf, int &$offset, int $count): array
    {
        $arr = [];
        for ($i = 0; $i < $count; $i++) {
            $arr[] = self::parseItem($buf, $offset);
        }
        return $arr;
    }
}

File: packages/core-ruby/lib/open_passkey/cbor_decoder.rb
# frozen_string_literal: true

module OpenPasskey
  module CborDecoder
    MAJOR_UNSIGNED_INT = 0
    MAJOR_NEGATIVE_INT = 1
    MAJOR_BYTE_STRING = 2
    MAJOR_TEXT_STRING = 3
    MAJOR_ARRAY = 4
    MAJOR_MAP = 5
    MAJOR_TAG = 6
    MAJOR_FLOAT_SIMPLE = 7

    class << self
      def decode(data)
        data = data.b if data.encoding != Encoding::ASCII_8BIT
        offset = [0]
        result = parse_item(data, offset)
        if offset[0] != data.bytesize
          raise WebAuthnError.new("cbor_error", "Unused bytes after data item")
        end
        result
      end

      def decode_in_place(data, start_offset)
        data = data.b if data.encoding != Encoding::ASCII_8BIT
        offset = [start_offset]
        result = parse_item(data, offset)
        [result, offset[0]]
      end

      private

      def parse_item(data, offset)
        first = read_byte(data, offset)
        type = first >> 5
        val = first & 0x1F

        if type == MAJOR_FLOAT_SIMPLE
          return parse_float_simple(val, data, offset)
        end

        val = parse_extra_length(val, data, offset)
        parse_item_data(type, val, data, offset)
      end

      def parse_float_simple(val, data, offset)
        case val
        when 20 then false
        when 21 then true
        when 22 then nil
        when 24
          v = read_byte(data, offset)
          parse_simple(v)
        when 25
          read_half_float(data, offset)
        when 26
          read_float(data, offset)
        when 27
          read_double(data, offset)
        when 28, 29, 30
          raise WebAuthnError.new("cbor_error", "Reserved value used")
        when 31
          raise WebAuthnError.new("cbor_error", "Indefinite length is not supported")
        else
          parse_simple(val)
        end
      end

      def parse_simple(val)
        case val
        when 20 then false
        when 21 then true
        when 22 then nil
        else
          raise WebAuthnError.new("cbor_error", "Unsupported simple value #{val}")
        end
      end

      def parse_extra_length(val, data, offset)
        case val
        when 0..23
          val
        when 24
          read_byte(data, offset)
        when 25
          read_uint16(data, offset)
        when 26
          read_uint32(data, offset)
        when 27
          read_uint64(data, offset)
        when 28, 29, 30
          raise WebAuthnError.new("cbor_error", "Reserved value used")
        when 31
          raise WebAuthnError.new("cbor_error", "Indefinite length is not supported")
        else
          val
        end
      end

      def parse_item_data(type, val, data, offset)
        case type
        when MAJOR_UNSIGNED_INT
          val
        when MAJOR_NEGATIVE_INT
          -1 - val
        when MAJOR_BYTE_STRING
          read_bytes(data, offset, val)
        when MAJOR_TEXT_STRING
          read_bytes(data, offset, val).force_encoding(Encoding::UTF_8)
        when MAJOR_ARRAY
          parse_array(data, offset, val)
        when MAJOR_MAP
          parse_map(data, offset, val)
        when MAJOR_TAG
          parse_item(data, offset)
        else
          raise WebAuthnError.new("cbor_error", "Unknown major type #{type}")
        end
      end

      def parse_map(data, offset, count)
        map = {}
        count.times do
          key = parse_item(data, offset)
          value = parse_item(data, offset)
          unless key.is_a?(Integer) || key.is_a?(String)
            raise WebAuthnError.new("cbor_error", "Map keys must be strings or integers")
          end
          map[key] = value
        end
        map
      end

      def parse_array(data, offset, count)
        arr = []
        count.times { arr << parse_item(data, offset) }
        arr
      end

      def read_byte(data, offset)
        raise WebAuthnError.new("cbor_error", "Read out of bounds") if offset[0] >= data.bytesize
        b = data.getbyte(offset[0])
        offset[0] += 1
        b
      end

      def read_bytes(data, offset, length)
        return "".b if length == 0
        if offset[0] + length > data.bytesize
          raise WebAuthnError.new("cbor_error", "Read out of bounds")
        end
        result = data.byteslice(offset[0], length)
        offset[0] += length
        result
      end

      def read_uint16(data, offset)
        bytes = read_bytes(data, offset, 2)
        bytes.unpack1("n")
      end

      def read_uint32(data, offset)
        bytes = read_bytes(data, offset, 4)
        bytes.unpack1("N")
      end

      def read_uint64(data, offset)
        bytes = read_bytes(data, offset, 8)
        bytes.unpack1("Q>")
      end

      def read_half_float(data, offset)
        half = read_uint16(data, offset)
        exp = (half >> 10) & 0x1F
        mant = half & 0x3FF
        sign = (half >> 15) != 0 ? -1.0 : 1.0

        if exp == 0
          sign * Math.ldexp(mant, -24)
        elsif exp == 31
          mant == 0 ? sign * Float::INFINITY : Float::NAN
        else
          sign * Math.ldexp(mant + 1024, exp - 25)
        end
      end

      def read_float(data, offset)
        bytes = read_bytes(data, offset, 4)
        bytes.unpack1("g")
      end

      def read_double(data, offset)
        bytes = read_bytes(data, offset, 8)
        bytes.unpack1("G")
      end
    end
  end
end
